### DISCUSSION TITLE
Hierarchical Tree Layout

### DISCUSSION THREAD

**Discussion by mlshv** [cosmograph-org/cosmos#110] [Q&A] [answered]
Hello!

I am impressed by Cosmograph's performance. It works much faster than any other tool I've tried for graph visualization

I am working on a social graph visualization project. It's a tree-shaped structure where nodes represent users, and they are connected to the users they invited. The graph isn't very large yet (around 20k nodes), but my current solution (Cytoscape.js) is lagging badly

However, I can't use Cosmograph because the layout doesn't suit my needs. I need a hierarchical tree layout with the root node at the top and child nodes below it. I understand this requires a different layout algorithm

Could you provide any guidance on whether this is possible with Cosmograph? If it’s feasible, I'd like to try forking Cosmograph and implementing it myself using an open-source tree layout algorithm as a reference

Thank you!

**Comment by rokotyan** [comment, marked as answer]
Hi @mlshv, thanks for the feedback!

Currently we're not planning to add new layout algorithms to Cosmograph, however you can pass your own layout as `x` and `y` properties of your node objects. Take a look at D3's tree layout or at Elkjs (https://github.com/kieler/elkjs). You can calculate your custom layout using one of these tools, and render your network with Cosmograph.

**Comment by jaads** [comment]
@rokotyan That likely does not solve the performance problem. 

I also use cytoscape.js and the dagre layout (which is baed on https://github.com/dagrejs/dagre). The layout calculation performs poorly for large graphs. That why I was also wondering too if cosmos.gl supports a tree/dagre layout. The layout calculation is currently a bottleneck for me, so I'd also be interested in a more performant layout calculation.
